src/test/java/org/openelisglobal/BaseWebContextSensitiveTest.java

```diff
@@ -44,4 +72,107 @@ public <T> T mapFromJson(String json, Class<T> clazz) throws IOException {
         objectMapper.enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY);
         return objectMapper.readValue(json, clazz);
     }
+
+    /**
+     * Executes a dataset with state management - preserves and restores the
+     * original state of affected tables after execution.
+     */
+    protected void executeDataSetWithStateManagement(String datasetFilename) throws Exception {
+        if (datasetFilename == null) {
+            throw new NullPointerException("Please provide test dataset file to execute!");
+        }
+
+        IDatabaseConnection connection = null;
+        try {
+            connection = new DatabaseConnection(dataSource.getConnection());
+            DatabaseConfig config = connection.getConfig();
+            config.setProperty(DatabaseConfig.FEATURE_ALLOW_EMPTY_FIELDS, true);
+
+            IDataSet newDataSet = loadDataSet(datasetFilename);
+            String[] tableNames = newDataSet.getTableNames();
+
+            // Backup current state of affected tables
+            IDataSet currentState = connection.createDataSet(tableNames);
+            originalStateCache.put(Arrays.toString(tableNames), currentState);
+            tablesToRestore.add(tableNames);
+
+            executeDataSet(datasetFilename);
+        } finally {
+            if (connection != null) {
+                connection.close();
+            }
+        }
+    }
+
+    /**
+     * This method will be called after each transaction to restore the database
+     * state
+     */
+    @AfterTransaction
+    @SuppressWarnings("unused")
+    protected void restoreDatabase() throws Exception {
```

Comment:
This method is automatically called after each transaction to restore the original state of the affected tables by using a backup. It employs `DatabaseOperation.CLEAN_INSERT` to reset the tables to their initial state and ensures that any changes made during the test are undone. After the restoration, it clears the state cache to maintain a clean state for subsequent tests.